foxpuppet/browser/window.py

```diff
@@ -2,43 +2,76 @@
 
+from foxpuppet.windows import BaseWindow
+
 
-class BrowserWindow(object):
+class BrowserWindow(BaseWindow):
     _file_menu_button_locator = (By.ID, 'file-menu')
     _file_menu_private_window_locator = (By.ID, 'menu_newPrivateWindow')
     _file_menu_new_window_button_locator = (By.ID, 'menu_newNavigator')
     _nav_bar_locator = (By.ID, 'nav-bar')
     _tab_browser_locator = (By.ID, 'tabbrowser-tabs')
 
     def __init__(self, selenium, *args, **kwargs):
+        super(BrowserWindow, self).__init__(selenium)
         self.selenium = selenium
-        self._windows = Windows(selenium)
 
     @property
     def is_private(self):
         """Returns True if this is a Private Browsing window."""
 
-        self.selenium.set_context('chrome')
-        return self.selenium.execute_script("""
+        with self.selenium.context('chrome'):
+            return self.selenium.execute_script(
+                """
                 Components.utils.import("resource://gre/modules/PrivateBrowsingUtils.jsm");
 
                 let chromeWindow = arguments[0].ownerDocument.defaultView;
                 return PrivateBrowsingUtils.isWindowPrivate(chromeWindow);
-            """, self._windows.window_element)
-        self.selenium.set_context('content')
+                """, self.window_element)
+
+    @property
+    def window_element(self):
+        """ Returns the inner DOM window element.
+
+        :returns: DOM window element.
+        """
+
+        return self.selenium.find_element(By.CSS_SELECTOR, ':root')
 
     def open_window(self, private=False):
-        self.selenium.set_context('chrome')
-        self.selenium.find_element(*self._file_menu_button_locator).click()
-        with self._windows.wait_for_new_window():
+        """Opens a new browser window
+
+        :param private: Optional parameter to open a private browsing window.
+                        Defaults to false.
+        """
+
+        with self.selenium.context('chrome'):
+            handles_before = self.selenium.window_handles
+
+            self.selenium.find_element(*self._file_menu_button_locator).click()
             if private:
                 self.selenium.find_element(
                     *self._file_menu_private_window_locator).click()
             else:
                 self.selenium.find_element(
                     *self._file_menu_new_window_button_locator).click()
-        self.selenium.set_context('content')
+        return self.wait.until(
+            new_window_is_opened(self.selenium, handles_before))
+
+
+class new_window_is_opened(object):
+    """
+    An expectation for checking that a new window is found after requesting
+    it to be opened
+    """
+
+    def __init__(self, selenium, handles_before):
+        self.selenium = selenium
+        self.handles_before = handles_before
+
+    def __call__(self, selenium):
+        for handle in self.selenium.window_handles:
+            if handle not in self.handles_before:
+                return handle
```

Comment:
Yeah its linear time. The more handles there are the longer it will take. I have to come up with something to verify that a new window has actually opened. This is dirty but it works for now I guess.